x-pack/filebeat/input/awscloudwatch/input.go

```diff
@@ -120,43 +119,67 @@ func (in *cloudwatchInput) Run(inputContext v2.Context, pipeline beat.Pipeline)
 	cwPoller := newCloudwatchPoller(
 		log.Named("cloudwatch_poller"),
 		in.metrics,
-		in.awsConfig.Region,
+		region,
 		in.config)
 	logProcessor := newLogProcessor(log.Named("log_processor"), in.metrics, client, ctx)
-	cwPoller.metrics.logGroupsTotal.Add(uint64(len(logGroupNames)))
+	cwPoller.metrics.logGroupsTotal.Add(uint64(len(logGroupIDs)))
 	cwPoller.startWorkers(ctx, svc, logProcessor)
-	cwPoller.receive(ctx, logGroupNames, time.Now)
+	cwPoller.receive(ctx, logGroupIDs, time.Now)
 	return nil
 }
 
-func parseARN(logGroupARN string) (string, string, error) {
-	arnParsed, err := arn.Parse(logGroupARN)
-	if err != nil {
-		return "", "", fmt.Errorf("error Parse arn %s: %w", logGroupARN, err)
-	}
+// fromConfig is a helper to parse input configurations and derive logGroupIDs & aws region
+// Returned logGroupIDs could be empty, which require other fallback mechanisms to derive them.
+// See getLogGroupNames for example.
+func fromConfig(cfg config, awsCfg awssdk.Config) (logGroupIDs []string, region string, err error) {
+	// LogGroupARN has precedence over LogGroupName & RegionName
+	if cfg.LogGroupARN != "" {
+		parsedArn, err := arn.Parse(cfg.LogGroupARN)
+		if err != nil {
+			return nil, "", fmt.Errorf("failed to parse log group ARN: %w", err)
+		}
 
-	if strings.Contains(arnParsed.Resource, ":") {
-		resourceARNSplit := strings.Split(arnParsed.Resource, ":")
-		if len(resourceARNSplit) >= 2 && resourceARNSplit[0] == "log-group" {
-			return resourceARNSplit[1], arnParsed.Region, nil
+		if parsedArn.Region == "" {
+			return nil, "", fmt.Errorf("failed to parse log group ARN: missing region")
 		}
+
+		// refine to match AWS API parameter regex of logGroupIdentifier
+		groupId := strings.TrimSuffix(cfg.LogGroupARN, ":*")
+		logGroupIDs = append(logGroupIDs, groupId)
+
+		return logGroupIDs, parsedArn.Region, nil
+	}
+
+	// then fallback to LogrGroupName
+	if cfg.LogGroupName != "" {
+		logGroupIDs = append(logGroupIDs, cfg.LogGroupName)
 	}
-	return "", "", fmt.Errorf("cannot get log group name from log group ARN: %s", logGroupARN)
+
+	// finally derive region
+	if cfg.RegionName != "" {
+		region = cfg.RegionName
+	} else {
+		region = awsCfg.Region
+	}
+
+	return logGroupIDs, region, nil
 }
 
-// getLogGroupNames uses DescribeLogGroups API to retrieve all log group names
-func getLogGroupNames(svc *cloudwatchlogs.Client, logGroupNamePrefix string, logGroupName string) ([]string, error) {
-	if logGroupNamePrefix == "" {
-		return []string{logGroupName}, nil
+// getLogGroupNames uses DescribeLogGroups API to retrieve LogGroupArn entries that matches the provided logGroupNamePrefix
+func getLogGroupNames(svc *cloudwatchlogs.Client, logGroupNamePrefix string, withLinkedAccount *bool) ([]string, error) {
+	var linkedAccounts bool
```

Comment:
If we change `IncludeLinkedAccountsForPrefixMode` in the type config struct in config.go to bool (instead of *bool), then we can remove line 170-172 and just pass `awssdk.Bool(withLinkedAccount)` into `DescribeLogGroupsInput`.

Reply:
I switched backed to non-pointer usage - https://github.com/elastic/beats/pull/41206/commits/7d37eb8d57192f4f875d8a432657e8f77c3d0f16

Default (if property not present is false)